File: src/main/kotlin/specscript/commands/types/CheckType.kt
package specscript.commands.schema

import specscript.language.*
import specscript.language.types.*
import specscript.util.toArrayNode
import specscript.util.toDomainObject
import tools.jackson.databind.JsonNode
import tools.jackson.databind.node.ObjectNode
import tools.jackson.databind.node.StringNode

object CheckType : CommandHandler("Check type", "core/schema"), ObjectHandler {

    override fun execute(data: ObjectNode, context: ScriptContext): JsonNode? {
        val json = data.getParameter("item")
        val typeData = data.getParameter("type")
        val type = getType(typeData, context)

        // Type validation throws exception when invalid
        validate(json, type, context.types)

        return null
    }
}

private fun validate(data: JsonNode, type: Type, registry: TypeRegistry) {

    val messages = type.definition.resolve(registry).definition.validate(data)

    if (messages.isNotEmpty()) {
        val validationErrors = messages.map { StringNode(it) }.toArrayNode()

        throw SpecScriptCommandError("Type validation errors", type = "Type validation", data = validationErrors)
    }
}

private fun getType(typeData: JsonNode, context: ScriptContext): Type {

    val type = typeData.toDomainObject(TypeSpecification::class)

    return type.resolve(context.types)
}

File: typescript/src/commands/check-type.ts
/**
 * Check type command — validates data against SpecScript type definitions.
 *
 * Mirrors Kotlin: specscript.commands.types.CheckType
 */

import type { CommandHandler } from '../language/command-handler.js'
import type { JsonValue } from '../language/types.js'
import { isObject, SpecScriptCommandError, CommandFormatError } from '../language/types.js'
import type { ScriptContext } from '../language/context.js'
import { parseTypeSpecification, resolveType, validateType } from '../language/type-system.js'

export const CheckTypeCommand: CommandHandler = {
  name: 'Check type',

  async execute(data: JsonValue, context: ScriptContext): Promise<JsonValue | undefined> {
    if (!isObject(data)) {
      throw new CommandFormatError('Check type requires an object with "item" and "type"')
    }

    const item = data['item']
    if (item === undefined) {
      throw new CommandFormatError('Check type requires an "item" property')
    }

    const typeData = data['type']
    if (typeData === undefined) {
      throw new CommandFormatError('Check type requires a "type" property')
    }

    // Parse and resolve the type specification
    const typeSpec = parseTypeSpecification(typeData)
    const resolvedType = resolveType(typeSpec, context.types)

    // Validate the data against the resolved type
    const messages = validateType(item, resolvedType.definition)
    if (messages.length > 0) {
      throw new SpecScriptCommandError(
        'Type validation errors',
        'Type validation',
        messages
      )
    }

    return undefined
  },
}
